File: LeetCode/0851.loud-and-rich/loud-and-rich.java
class Solution {
    public int[] loudAndRich(int[][] richer, int[] quiet) {
        int[] res = new int[quiet.length];
        Arrays.fill(res,-1);
        HashMap<Integer,List<Integer>>mp = new HashMap<>();
        for (int i = 0;i < quiet.length;i++){
            mp.put(i,new ArrayList<>());
        }
        for (int[] per : richer){
            mp.get(per[1]).add(per[0]);
        }
        for (int i = 0;i < quiet.length;i++){
            dfs(mp,i,quiet,res);
        }
        return res;
    }

    public void dfs(HashMap<Integer,List<Integer>>mp,int index,int[] quiet,int[] res){
        if (res[index] >= 0)    return;
        res[index] = index;
        for (int val : mp.get(index)){
            dfs(mp,val,quiet,res);
            if (quiet[res[val]] < quiet[res[index]]){
                res[index] = res[val];
            }
        }
    }
}

File: LeetCode/0851.loud-and-rich/loud-and-rich.cpp
class Solution {
public:
    vector<int> loudAndRich(std::vector<std::vector<int>>& richer,vector<int>& quiet) {
		vector<int> result(quiet.size(), -1);

		unordered_map<int, vector<int>>graph;

		for (const auto& vec : richer) {
			graph[vec[1]].push_back(vec[0]);
		}

		for (int i = 0; i < result.size(); ++i) {
			helper(graph, i, quiet, result);
		}

		return result;

	}


	void helper(unordered_map<int, vector<int>>& graph, int index, const vector<int>& quiet, vector<int>& result) {
		
	    if (result[index] >= 0) {
			return;
		}
		
		result[index] = index;
		for (auto& element_index : graph[index]) {
			helper(graph, element_index, quiet, result);
			
			if (quiet[result[element_index]] < quiet[result[index]]) {
				result[index] = result[element_index];
			}
		}
	}
};